File: uhr/cpp/src/logging.cc
#include "logging.h"
#include "util.h"

namespace uhr {

	LogSink::Acquired::~Acquired() {
		if (sink_->callback_ != nullptr) {
			auto s = ToUTF16(oss_.str());
			auto ref = ToStringRef(s);
			sink_->callback_(ref, sink_->user_data_);
		}
	}

	LogSink::Acquired::Acquired(LogSink *sink) : sink_(sink), lock_(sink->mutex_) {}

	LogSink::Acquired LogSink::Acquire() {
		return {this};
	}

	void LogSink::Set(UHR_LoggingCallback callback, void* user_data) {
		std::lock_guard<std::mutex> lock(mutex_);
		user_data_ = user_data;
		callback_ = callback;
	}

	// Not inside anonymous namespace, so it can be accessed by the clue
	// logging macros in other code units.
	LogSink g_log_sink;

} // namespace uhr

File: uhr/objc/src/Logging.m
#import "Logging.h"

LogSink* gLogSink = [[LogSink alloc] init];

@interface LogSink()
@property (nonatomic) UHR_LoggingCallback callback;
@property (nonatomic) void* userData;
@end

@implementation LogSink

- (id)init {
    self = [super init];
    if (self) {
        self.callback = nil;
        self.userData = nil;
    }
    return self;
}

- (void)set:(UHR_LoggingCallback)callback userData:(void *)userData {
    @synchronized(self) {
        self.callback = callback;
        self.userData = userData;
    }
}

- (void)log:(NSString*)str {
    @synchronized(self) {
        if (self.callback != nil) {
            UHR_StringRef ref;
            ref.characters = (const uint16_t*)CFStringGetCharactersPtr((__bridge CFStringRef)str);
            ref.length = (uint32_t)str.length;
            self.callback(ref, self._userData);
        }
    }
}

- (void)dealloc {
    [super dealloc];
}

@end
